### PR TITLE
Possibly about a run-time configuration, or maybe just a Deployment

### PR DESCRIPTION
**Pull request by jboxman**
So this presents a problem, as we don't actually talk about a run-time configuration for SR-IOV or Multus anywhere, really, ever.

Preview: https://bz-1846333--ocpdocs.netlify.app/openshift-enterprise/latest/networking/hardware_networks/add-pod.html (for 4.6)

### CODE REVIEW

**Comment by jboxman** [comment]
@zshi-redhat, this reflects my current thinking on this. But a larger discussion of `k8s.v1.cni.cncf.io/networks` probably needs to happen elsewhere in the documentation.

**Comment by zshi-redhat** [comment]
> @zshi-redhat, this reflects my current thinking on this. But a larger discussion of `k8s.v1.cni.cncf.io/networks` probably needs to happen elsewhere in the documentation.

@jboxman is this PR meant to cover this one: https://github.com/openshift/openshift-docs/pull/23382 ?

**Comment by jboxman** [comment]
@zshi-redhat, it's meant to cover the BZ, which as I understand it requires `namespace` in certain situations, so I thought it best to simply include `namespace` as a required field.

As long as `namespace` is properly specified, can't someone use a Pod definition in any k8s or OCP workload, whether it is a Deployment or DaemonSet or ReplicationController or whatever? I want to avoid introducing new procedures when clarifying an existing configuration is enough.

**Comment by zshi-redhat** [comment]
> @zshi-redhat, it's meant to cover the BZ, which as I understand it requires `namespace` in certain situations, so I thought it best to simply include `namespace` as a required field.
> 
> As long as `namespace` is properly specified, can't someone use a Pod definition in any k8s or OCP workload, whether it is a Deployment or DaemonSet or ReplicationController or whatever? I want to avoid introducing new procedures when clarifying an existing configuration is enough.

It will be slightly different in terms of using namespace for other multus pods and sriov pods.
For other multus pods, it's preferrable to NOT use `namespace` field as a separate option, instead, using the format `<namespace>/<name>`.
For sriov pods, since there is an issue with network-resources-injector that is not able to recognize/deduce the correct namespace for Deployment/DaemonSet/ReplicationController/StatefulSet, so it has to specify the namespace as a separate option in pod annotation. (I'm working on a fix for this issue, by the time the issue is fixed, we should be able to remove the namespace here)

**Comment by jboxman** [comment]
@zshi-redhat, then perhaps I am no closer to a solution.

- How does someone specify a namespace for a multus pod? Is this always necessary?
- How does someone specify a namespace for an SR-IOV pod? Is this always necessary?

The procedure should reflect the proper way to do this for each, and only the correct procedure will be shown based on where the content is included: The hardware networks section or the multiple networks section.

Ideally, a configuration that is valid for a pod `spec` is _also_ valid anywhere a pod template is valid; and that's why I do not cover workload configurations anywhere. But that could be an incorrect assumption?

**Comment by zshi-redhat** [comment]
> @zshi-redhat, then perhaps I am no closer to a solution.
> 
>     * How does someone specify a namespace for a multus pod? Is this always necessary?
> 
>     * How does someone specify a namespace for an SR-IOV pod? Is this always necessary?
> 

There was an [issue](https://bugzilla.redhat.com/show_bug.cgi?id=1846333) with sriov admission controller which requires the namespace field be specified when running a deploymen-like resource. The issue was fixed few days ago, I think we can consider that multus & sriov pod should use the same namespace format now.

File: modules/nw-multus-add-pod.adoc [hunk 1]
```diff
@@ -13,11 +13,18 @@ endif::[]
...
 ifdef::sriov[]
 [NOTE]
 =====
-The Network Resources Injector will inject the `resource` parameter into the Pod CR automatically if a NetworkAttachmentDefinition CR associated with the SR-IOV CNI plug-in is specified.
+The Network Resources Injector will inject the `resource` field into the Pod custom resource (CR) automatically if a NetworkAttachmentDefinition CR associated with the SR-IOV Container Network Interface (CNI) plug-in is specified.
 =====
+
+[IMPORTANT]
+====
+When specifying an SR-IOV hardware network, you must specify the namespace of the NetworkAttachmentDefinition CR. For more information, see the following bugs: link:https://bugzilla.redhat.com/show_bug.cgi?id=1846333[BZ#1846333] and link:https://bugzilla.redhat.com/show_bug.cgi?id=1840962[BZ#1840962].
```
Comment (jboxman) [inline modules/nw-multus-add-pod.adoc:26]: @zshi-redhat, so I can remove this reference to these bugs, then?

Comment (zshi-redhat) [reply modules/nw-multus-add-pod.adoc:26]: These two bugs were fixed in 4.6, we should remove them for 4.6 doc.
It should be added back when backporting to 4.5 doc or earlier releases (because the bug fix is not cherry-pick to 4.5 or earlier releases).

Comment (jboxman) [reply modules/nw-multus-add-pod.adoc:37]: @zshi-redhat, okay, I've added a conditional so that these bugs appear only in 4.3, 4.4, and 4.5.

Comment (zshi-redhat) [reply modules/nw-multus-add-pod.adoc:37]: @jboxman the fix looks good to me.

nit: since the issue only happens when sriov network is specified in a deployment/replicacontroller, shall we 
change:
`when specifying an SR-IOV hardware network`
to:
`when specifying an SR-IOV hardware network for a deployment or replicacontroller` ?

Comment (jboxman) [reply modules/nw-multus-add-pod.adoc]: @zshi-redhat, updated, thanks!

File: modules/nw-multus-add-pod.adoc [hunk 2]
```diff
@@ -7,17 +7,36 @@ ifeval::["{context}" == "configuring-sr-iov"]
...
+:bz:
+endif::[]
+
 [id="nw-multus-add-pod_{context}"]
 = Adding a Pod to an additional network
 
 You can add a Pod to an additional network. The Pod continues to send normal
 cluster related network traffic over the default network.
 
+If a Pod already exists, you cannot attach it to an additional network. Only when a Pod is created are any additional networks requested attached to the Pod.
```
Comment (bobfuru) [inline modules/nw-multus-add-pod.adoc:26]: To offset the first sentence, it might help to start the sentence, "However, if a Pod already exists...." 

Also, the next sentence is a bit hard to follow who is doing what to whom. If it's still accurate, maybe something like, "Additional requested networks are only attached when a Pod is created."

File: modules/nw-multus-add-pod.adoc [hunk 3]
```diff
@@ -7,17 +7,36 @@ ifeval::["{context}" == "configuring-sr-iov"]
...
+endif::[]
+ifeval::["{product-version}" == "4.5"]
+:bz:
+endif::[]
+
 [id="nw-multus-add-pod_{context}"]
 = Adding a Pod to an additional network
 
 You can add a Pod to an additional network. The Pod continues to send normal
 cluster related network traffic over the default network.
```
Comment (bobfuru) [inline modules/nw-multus-add-pod.adoc:24]: s/cluster related network/cluster-related network/

File: modules/nw-multus-add-pod.adoc [hunk 4]
```diff
@@ -7,17 +7,36 @@ ifeval::["{context}" == "configuring-sr-iov"]
...
 
+If a Pod already exists, you cannot attach it to an additional network. Only when a Pod is created are any additional networks requested attached to the Pod.
+
 ifdef::sriov[]
 [NOTE]
 =====
-The Network Resources Injector will inject the `resource` parameter into the Pod CR automatically if a NetworkAttachmentDefinition CR associated with the SR-IOV CNI plug-in is specified.
+The Network Resources Injector will inject the `resource` field into the Pod custom resource (CR) automatically if a NetworkAttachmentDefinition CR associated with the SR-IOV Container Network Interface (CNI) plug-in is specified.
```
Comment (bobfuru) [inline modules/nw-multus-add-pod.adoc:31]: s/will inject/injects/
Also, I question whether the word _automatically_ is necessary.
And is it still accurate to word it as follows?

> The Network Resources Injector injects the `resource` field into the Pod custom resource (CR) if the SR-IOV Container Network Interface (CNI) plug-in specifies a NetworkAttachmentDefinition CR.

Trying to clarify for the reader what is meant by "is specified".

File: modules/nw-multus-add-pod.adoc [hunk 5]
```diff
@@ -31,57 +50,57 @@ endif::sriov[]
...
-<1> Replace `<network>` with the name of the additional network to associate
-with the Pod. To specify more than one additional network, separate each network
+<1> To specify more than one additional network, separate each network
 with a comma. Do not include whitespace between the comma. If you specify
 the same additional network multiple times, that Pod will have multiple network
 interfaces attached to that network.
-+
-In the following example, two additional networks are attached to the Pod:
+
+.. To attach an additional network with customizations, add an annotation with the following format.
```
Comment (bobfuru) [inline modules/nw-multus-add-pod.adoc:68]: Same comment about ending with `:`.

File: modules/nw-multus-add-pod.adoc [hunk 6]
```diff
@@ -31,57 +50,57 @@ endif::sriov[]
...
-    command: ["/bin/bash", "-c", "sleep 2000000000000"]
-    image: centos/tools
+    k8s.v1.cni.cncf.io/networks: |-
+      [
+        "name": "<network>", <1>
+        "namespace": "<namespace>", <2>
+        "default-route": ["<default-route>"] <3>
+      ]
 ----
+<1> Specify the name of the additional network defined by an NetworkAttachmentDefinition CR.
```
Comment (bobfuru) [inline modules/nw-multus-add-pod.adoc:81]: s/an NetworkAttachmentDefinition/a NetworkAttachmentDefinition/

File: modules/nw-multus-add-pod.adoc [hunk 7]
```diff
@@ -31,57 +50,57 @@ endif::sriov[]
...
+. To create the Pod, enter the following command:
 +
+[source,terminal]
 ----
 $ oc create -f pod.yaml
 ----
 
-. Optional: Confirm that the annotation exists in the Pod CR by running the
-following command. Replace `<name>` with the name of the Pod.
+. Optional: To Confirm that the annotation exists in the Pod CR, enter the following command. Replace `<name>` with the name of the Pod.
```
Comment (bobfuru) [inline modules/nw-multus-add-pod.adoc:93]: Might help readability as:
_...enter the following command, replacing `<name>` with..._

**Review by bobfuru — COMMENTED** [review]
Couple of questions and a few alternate wording suggestions but otherwise this LGTM.

**Comment by jboxman** [comment]
@bobfuru thanks!

File: modules/nw-multus-add-pod.adoc (hunk 2 above)
Comment (jboxman) [reply modules/nw-multus-add-pod.adoc:26]: @bobfuru, true, it doesn't really make that much sense as it is.

Comment (jboxman) [reply modules/nw-multus-add-pod.adoc:26]: @bobfuru might try

> When a Pod is created additional networks are attached to it. However, if a Pod already exists, you cannot attach additional networks to it.

File: modules/nw-multus-add-pod.adoc (hunk 4 above)
Comment (jboxman) [reply modules/nw-multus-add-pod.adoc:31]: @bobfuru, unrelated, but as it happens Pod is part of k8s, so I am incorrect when referring to it as a CR. A Pod is just a resource or, like anything else, an object.

Comment (jboxman) [reply modules/nw-multus-add-pod.adoc:31]: @bobfuru, I think the goal is to explain this:

> If a NetworkAttachmentDefinition is managed by the SR-IOV Network Operator, the SR-IOV Network Resource Injector adds the `resource` field to the Pod object automatically.

At some point, this seemed important, probably because it was in the original contributed content. Maybe it's helpful when someone sees the field and doesn't know why it is there.

File: modules/nw-multus-add-pod.adoc [hunk 8]
```diff
@@ -31,57 +50,57 @@ endif::sriov[]
 
 .Procedure
 
-To add a Pod with additional networks, complete the following steps:
+. Add an annotation to the Pod resource definition. Only one of the following annotation formats can be used.
```
[the remark this answers is not in the document]
Comment (jboxman) [reply modules/nw-multus-add-pod.adoc:53]: @bobfuru, I haven't happened upon a suitable approach still. But a Pod is never a CR I've discovered. (All of OpenShift is implemented with CRs, so it at first appears that it's correct always, but apparently not.)

I wonder if we'll settle our terminology soon?

File: modules/nw-multus-add-pod.adoc [hunk 9]
```diff
@@ -31,57 +50,57 @@ endif::sriov[]
...
 
-. Create the Pod resource definition and add the `k8s.v1.cni.cncf.io/networks`
-parameter to the Pod `metadata` mapping. The `k8s.v1.cni.cncf.io/networks`
-accepts a comma separated string of one or more NetworkAttachmentDefinition Custom Resource (CR) names:
+.. To attach an additional network without any customization, add an annotation with the following format. Replace `<network>` with the name of the additional network to associate with the Pod.
```
[the remark this answers is not in the document]
Comment (jboxman) [reply modules/nw-multus-add-pod.adoc:55]: @bobfuru, interestingly, I too have seen both ways. I've been undoing all my note annotations, and trying to keep the number in the preceding text to a minimum as well.

Lately I think if a procedure has a heavily annotated object or configuration, it needs to be explained in a different section. If only we could use `xref` in modules. (SR-IOV is particularly bad about this, as I took the bulk of the content, tried to clean it up, and merged it. But it is all procedures with heavily annotated object definitions.)

In any case, I don't think we have a documented standard for this.

File: modules/nw-multus-add-pod.adoc (hunk 7 above)
Comment (jboxman) [reply modules/nw-multus-add-pod.adoc:93]: @bobfuru, but does it feel more like a run-on sentence? I wanted to avoid abusing commas.

**Comment by jboxman** [comment]
@zshi-redhat, so in 4.6+, would anyone ever want to specify `namespace` in:

```
metadata:
  annotations:
    k8s.v1.cni.cncf.io/networks: |-
      [
        "name": "<network>", <1>
        "namespace": "<namespace>", <2>
        "default-route": ["<default-route>"] <3>
      ]
```

Or can I omit `namespace` completely for SR-IOV and non-SR-IOV additional networks? Or in 4.6+ is it _optional_ to specify a namespace?

Thanks!

File: modules/nw-multus-add-pod.adoc (hunk 7 above)
Comment (bobfuru) [reply modules/nw-multus-add-pod.adoc]: Fair point!

File: modules/nw-multus-add-pod.adoc (hunk 9 above)
[the remark this answers is not in the document]
Comment (bobfuru) [reply modules/nw-multus-add-pod.adoc]: Case-by-case basis seems reasonable then, while we pine away for xrefs.

**Comment by zshi-redhat** [comment]
> @zshi-redhat, so in 4.6+, would anyone ever want to specify `namespace` in:
> 
> ```
> metadata:
>   annotations:
>     k8s.v1.cni.cncf.io/networks: |-
>       [
>         "name": "<network>", <1>
>         "namespace": "<namespace>", <2>
>         "default-route": ["<default-route>"] <3>
>       ]
> ```
> 
> Or can I omit `namespace` completely for SR-IOV and non-SR-IOV additional networks? Or in 4.6+ is it _optional_ to specify a namespace?

For SR-IOV, we fixed the issue where namespace has to be explicitly specified even for deployment/replica pod that uses the net-attach-def in the same namespace. For non-SRIOV pod, we don't have such issue.
So for creating a pod that refers to the net-attach-def in the same namespace, we don't need to specify the `namespace` field in pod annotation.

I remembered there is a namespace isolation feature in multus, @dougbtv may help comment on whether namespace is needed with that feature.

**Comment by dougbtv** [comment]
So.... there is one scenario where using namespace is beneficial, and can be used in two different ways... However, I'm not sure if it's the run-time specific usage where it shows benefit.

Usually, it'd be used this way in the CSV list format:

```
k8s.v1.cni.cncf.io/networks: default/my-net-attach-def
```

or, can be used here, too, with: 

```
k8s.v1.cni.cncf.io/networks: |-
      [{
        "name": "my-net-attach-def", 
        "namespace": "default"
      }]
```

Especially important here is the use of the `default` namespace. We do use a feature in OCP that's called "namespace isolation" as Zenghui noted. This prevents you from referring to namespaces for net-attach-defs outside of the namespace of a given pod. Which is a security feature, or a way to help you define roles for users, really.

However, the `default` namespace is special. You can re-use net-attach-defs in the default namespace from any namespace in the cluster. To prevent you from having to create the same net-attach-def in a bunch of different namespaces.

...This was new in 4.5 IIRC.

Also while I'm here... I see Zenghui's example has square braces for a list but no curly braces for objects.... In the spec in section: `4.1.2 JSON List Format` it should be a list of JSON objects, so, something like what I have above, or with multiple items:

```
k8s.v1.cni.cncf.io/networks: |-
      [{
        "name": "my-net-attach-def", 
        "namespace": "default"
      },{
        "name": "another-thing"
      }]
```
